File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.py
from collections import deque
def solution(players, m, k):
    cnt = 0
    live_server = deque()
    for i in range(len(players)):
        while live_server and live_server[0] == i:
            live_server.popleft()
        diff = players[i]//m - len(live_server)
        if diff>0:
            cnt+=diff
            for _ in range(diff):
                live_server.append(i+k)
        
    return cnt

File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.java
import java.util.Deque;
import java.util.LinkedList;

public class Solution {
    public int solution(int[] players, int m, int k) {
        int cnt = 0;
        Deque<Integer> liveServer = new LinkedList<>();
        
        for (int i = 0; i < players.length; i++) {
            while (!liveServer.isEmpty() && liveServer.peekFirst() == i) {
                liveServer.pollFirst();
            }
            
            int diff = (players[i] / m) - liveServer.size();
            
            if (diff > 0) {
                cnt += diff;
                for (int j = 0; j < diff; j++) {
                    liveServer.add(i + k);
                }
            }
        }
        
        return cnt;
    }
}
